mldsa/sign.c

```diff
@@ -204,74 +325,36 @@ int crypto_sign_signature_internal(uint8_t *sig, size_t *siglen,
   /* place loop invariants for CBMC.                          */
   while (1)
   __loop__(
-    invariant(1) /* TODO */
+    assigns(nonce, object_whole(siglen), memory_slice(sig, CRYPTO_BYTES))
+    invariant(nonce <= NONCE_UB)
+
+    /* t0, s1, s2, and mat are initialized above and are NOT changed by this */
+    /* loop. We can therefore re-assert their bounds here as part of the     */
+    /* loop invariant. This makes proof noticeably faster with CBMC          */
+    invariant(forall(k1, 0, MLDSA_K, forall(l1, 0, MLDSA_L,
+              array_bound(mat[k1].vec[l1].coeffs, 0, MLDSA_N, 0, MLDSA_Q))))
+    invariant(forall(k2, 0, MLDSA_K, array_abs_bound(t0.vec[k2].coeffs, 0, MLDSA_N, MLD_NTT_BOUND)))
+    invariant(forall(k3, 0, MLDSA_L, array_abs_bound(s1.vec[k3].coeffs, 0, MLDSA_N, MLD_NTT_BOUND)))
+    invariant(forall(k4, 0, MLDSA_K, array_abs_bound(s2.vec[k4].coeffs, 0, MLDSA_N, MLD_NTT_BOUND)))
   )
   {
-    /* Sample intermediate vector y */
-    mld_polyvecl_uniform_gamma1(&y, rhoprime, nonce++);
-
-    /* Matrix-vector multiplication */
-    z = y;
-    mld_polyvecl_ntt(&z);
-    mld_polyvec_matrix_pointwise_montgomery(&w1, mat, &z);
-    mld_polyveck_reduce(&w1);
-    mld_polyveck_invntt_tomont(&w1);
-
-    /* Decompose w and call the random oracle */
-    mld_polyveck_caddq(&w1);
-    mld_polyveck_decompose(&w1, &w0, &w1);
-    mld_polyveck_pack_w1(sig, &w1);
-
-    mld_H(sig, MLDSA_CTILDEBYTES, mu, MLDSA_CRHBYTES, sig,
-          MLDSA_K * MLDSA_POLYW1_PACKEDBYTES, NULL, 0);
-    mld_poly_challenge(&cp, sig);
-    mld_poly_ntt(&cp);
-
-    /* Compute z, reject if it reveals secret */
-    mld_polyvecl_pointwise_poly_montgomery(&z, &cp, &s1);
-    mld_polyvecl_invntt_tomont(&z);
-    mld_polyvecl_add(&z, &y);
-    mld_polyvecl_reduce(&z);
-    if (mld_polyvecl_chknorm(&z, MLDSA_GAMMA1 - MLDSA_BETA))
-    {
-      /* reject */
-      continue;
-    }
+    int result;
 
-    /* Check that subtracting cs2 does not change high bits of w and low bits
-     * do not reveal secret information */
-    mld_polyveck_pointwise_poly_montgomery(&h, &cp, &s2);
-    mld_polyveck_invntt_tomont(&h);
-    mld_polyveck_sub(&w0, &h);
-    mld_polyveck_reduce(&w0);
-    if (mld_polyveck_chknorm(&w0, MLDSA_GAMMA2 - MLDSA_BETA))
+    if (nonce == NONCE_UB)
```

Comment:
Please add a comment what is going on here and that this should never happen - also describe how it differs from the reference implementation.